src/lib.rs

```diff
@@ -1,564 +1,570 @@
+        let sequences = reader.sequence_records()?;
+        Ok(Self {
+            reader,
+            sequences,
+            softmask_enabled: false,
         })
     }
 
-    /// Get all hard blocks (N-blocks) of a chromosome
+    /// Reads a partial sequence of a chromosome.
     ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn full_hard_masked_blocks(&self, chr: &str) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => Ok(record.n_blocks),
-            Err(e) => Err(e),
-        }
-    }
+    /// * `chr` – name of the chromosome
+    /// * `range` – selected range of the sequence (`..` for full sequence)
+    pub fn read_sequence(
+        &mut self,
+        chr: impl AsRef<str>,
+        range: impl RangeBounds<usize>,
+    ) -> Result<String> {
+        const NUC: &[u8; 4] = b"TCAG";
+
+        let seq = self.sequences.query(chr)?;
+        let reader = &mut self.reader;
+
+        let start = match range.start_bound() {
+            Bound::Included(&v) => v,
+            Bound::Excluded(&v) => v + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&v) => v + 1,
+            Bound::Excluded(&v) => v,
+            Bound::Unbounded => seq.length,
+        };
 
-    /// Get hard blocks (N-blocks) of a region on a chromosome
-    ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn hard_masked_blocks(
-        &self,
-        chr: &str,
-        start: usize,
-        end: usize,
-    ) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => {
-                let mut result = Vec::new();
-                for block in record.n_blocks {
-                    let block_end = block.start + block.length;
-                    if block_end as usize <= start {
-                        continue;
-                    } else if block.start as usize > end {
-                        break;
-                    }
-                    result.push(block)
-                }
-                Ok(result)
-            }
-            Err(e) => Err(e),
+        let first_byte = start / 4;
+        reader.seek(SeekFrom::Start(seq.offset))?; // beginning of the DNA sequence
+        reader.seek(SeekFrom::Current(first_byte as _))?; // position where we want to start reading
+        if start >= end {
+            return Ok(String::new()); // trivial case, empty return result
         }
-    }
 
-    /// Get all soft blocks (lower case blocks) of a chromosome
-    ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn full_soft_masked_blocks(&self, chr: &str) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => Ok(record.soft_mask_blocks),
-            Err(e) => Err(e),
+        let length = end - start;
+        let mut out = Vec::with_capacity(length);
+        unsafe {
```

Comment:
Needs a comment above the unsafe block.

Reply:
Maybe we can actually use an array instead of a vec here? Then the size of the array would be static.

Reply:
We can't use an array here because length is not known at compile time

Reply:
Makes sense. But maybe we could do this instead?

`let mut out = vec![0; length];`

I created a simple Rust program that included that line and I looked at the generated assembly code. The code calls `memset` which is very efficient at initializing a region of memory.

Reply:
It's not a big deal here, but I would argue that simple `malloc` is still a bit faster - why set all those zeroes when we don't need it as they will get overwritten a few lines below? (especially when the sequence is long)

`set_len()` is unsafe in two cases:
- capacity is smaller than length
- objects inside the vector require some invariants to hold, or are non-trivial

Here, none of the above is a problem - we just allocated enough capacity for resizing, and there are no invariants attached to simple integers, whatever they're equal to is fine and they will get overwritten anyway.